rust/lance-index/src/scalar/bitmap.rs

```diff
@@ -323,45 +339,95 @@ fn get_batch_from_arrays(
     Ok(RecordBatch::try_new(schema, columns)?)
 }
 
-// Takes an iterator of Vec<u64> and processes each vector
-// to turn it into a RoaringTreemap. Each RoaringTreeMap is
-// serialized to bytes. The entire collection is converted to a BinaryArray
-fn get_bitmaps_from_iter<I>(iter: I) -> Arc<dyn Array>
-where
-    I: Iterator<Item = RowIdTreeMap>,
-{
-    let mut builder = BinaryBuilder::new();
-    iter.for_each(|bitmap| {
+fn chunked_keys_and_bitmaps(
+    state: HashMap<ScalarValue, RowIdTreeMap>,
+    value_type: &DataType,
+) -> Vec<(Arc<dyn Array>, Arc<dyn Array>)> {
+    let mut batches = Vec::new();
+    let mut cur_keys = Vec::new();
+    let mut cur_bitmaps = Vec::new();
+    let mut cur_bytes = 0;
+
+    for (key, bitmap) in state.into_iter() {
         let mut bytes = Vec::new();
         bitmap.serialize_into(&mut bytes).unwrap();
-        builder.append_value(&bytes);
-    });
+        let bitmap_len = bytes.len();
+
+        // If adding this bitmap would overflow, flush current batch
+        if cur_keys.len() >= MAX_BITMAP_ARRAY_LENGTH
+            || cur_bytes + bitmap_len > MAX_BITMAP_ARRAY_LENGTH
```

Comment:
I suppose it doesn't hurt but the check on `cur_key.len` seems superfluous, `cur_bytes` should always be larger than `cur_keys.len`